File: backend/app/routes/gift.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import Dict, Any, List
from datetime import datetime

from app.database import get_db
from app.models.user_models import User
from app.schemas.gift_schemas import GiftRequest, GiftActionRequest, GiftStatus as GiftStatusSchema
from app.services.gift_service import GiftService
from app.services.auth import get_current_user_from_token as get_current_user

router = APIRouter(prefix="/gift", tags=["gift"])


@router.get("/inbox", response_model=Dict[str, Any])
def get_gift_inbox_endpoint(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Vue consolidée des cadeaux reçus/envoyés avec métriques live."""
    try:
        gift_service = GiftService(db)
        return gift_service.get_gift_inbox(current_user.id)
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur lors de la récupération de la boîte aux cadeaux: {str(e)}"
        )


@router.get("/history", response_model=List[Dict[str, Any]])
def get_gift_history_endpoint(
    gift_type: str = "received",
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Récupérer l'historique des cadeaux (nouveau format étendu)
    Supporte à la fois legacy et new flow
    """
    try:
        gift_service = GiftService(db)
        gifts = gift_service.get_gift_history(current_user.id, gift_type)
        return gifts
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur lors de la récupération de l'historique: {str(e)}"
        )


@router.post("/send", response_model=Dict[str, Any], status_code=status.HTTP_201_CREATED)
def send_gift_endpoint(
    gift_data: GiftRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Envoyer un cadeau - NOUVEAU FLOW ATOMIQUE
    Utilise le flow: CREATED → PAID → DELIVERED
    """
    try:
        gift_service = GiftService(db)
        
        # Préparer les données pour le service
        gift_dict = {
            'receiver_phone': gift_data.receiver_phone,
            'bom_id': gift_data.bom_id,
            'message': gift_data.message,
            'quantity': gift_data.quantity
        }
        
        # Appeler le service atomique
        result = gift_service.send_gift(current_user.id, gift_dict)
        
        if not result.get("success"):
            raise HTTPException(
                status_code=400,
                detail=result.get("message", "Erreur lors de l'envoi du cadeau")
            )
        
        # Retourner le résultat complet du service
        return {
            "success": True,
            "message": result.get("message", "🎁 Cadeau envoyé avec succès!"),
            "gift_id": result.get("gift_id"),
            "transaction_reference": result.get("transaction_reference"),
            "financial": result.get("financial", {}),
            "timestamps": result.get("timestamps", {}),
            "status": "DELIVERED"
        }
        
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur interne lors de l'envoi du cadeau: {str(e)}"
        )


@router.post("/accept", response_model=Dict[str, Any])
def accept_gift_endpoint(
    gift_action: GiftActionRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Accepter un cadeau reçu - LEGACY FLOW ONLY
    Pour les cadeaux créés avant la migration (status SENT)
    """
    try:
        if gift_action.action != GiftStatusSchema.ACCEPTED:
            raise HTTPException(
                status_code=400,
                detail="Action non valide. Utilisez 'accepted' pour accepter un cadeau."
            )
        
        gift_service = GiftService(db)
        result = gift_service.accept_gift(gift_action.gift_id, current_user.id)
        
        if not result.get("success"):
            raise HTTPException(
                status_code=400,
                detail=result.get("message", "Erreur lors de l'acceptation du cadeau")
            )
        
        return result
        
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur lors de l'acceptation du cadeau: {str(e)}"
        )


@router.post("/decline", response_model=Dict[str, Any])
def decline_gift_endpoint(
    gift_action: GiftActionRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Refuser un cadeau reçu - LEGACY FLOW ONLY
    Pour les cadeaux créés avant la migration (status SENT)
    """
    try:
        if gift_action.action != GiftStatusSchema.DECLINED:
            raise HTTPException(
                status_code=400,
                detail="Action non valide. Utilisez 'declined' pour refuser un cadeau."
            )
        
        gift_service = GiftService(db)
        result = gift_service.decline_gift(gift_action.gift_id, current_user.id)
        
        if not result.get("success"):
            raise HTTPException(
                status_code=400,
                detail=result.get("message", "Erreur lors du refus du cadeau")
            )
        
        return result
        
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur lors du refus du cadeau: {str(e)}"
        )


@router.get("/pending", response_model=List[Dict[str, Any]])
def get_pending_gifts_endpoint(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Récupérer les cadeaux en attente - LEGACY FLOW ONLY
    Pour les cadeaux SENT qui n'ont pas encore été acceptés/refusés
    """
    try:
        gift_service = GiftService(db)
        pending_gifts = gift_service.get_pending_gifts(current_user.id)
        return pending_gifts
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur récupération cadeaux en attente: {str(e)}"
        )


@router.post("/expire-old", response_model=Dict[str, Any])
def expire_old_gifts_endpoint(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Nettoyer les cadeaux expirés (pour tâche background/admin)
    """
    try:
        # Vérifier les permissions si nécessaire
        # (optionnel: ajouter une vérification de rôle admin)
        
        gift_service = GiftService(db)
        expired_count = gift_service.expire_old_gifts()
        
        return {
            "success": True,
            "message": f"{expired_count} cadeaux expirés/nettoyés",
            "expired_count": expired_count,
            "timestamp": datetime.utcnow().isoformat()
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur lors du nettoyage des cadeaux expirés: {str(e)}"
        )


@router.get("/{gift_id}/details", response_model=Dict[str, Any])
def get_gift_details_endpoint(
    gift_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Récupérer les détails complets d'un cadeau spécifique
    """
    try:
        from app.models.gift_models import GiftTransaction
        from app.models.bom_models import UserBom
        
        # Récupérer le cadeau
        gift = db.query(GiftTransaction).filter(
            GiftTransaction.id == gift_id
        ).first()
        
        if not gift:
            raise HTTPException(
                status_code=404,
                detail="Cadeau non trouvé"
            )
        
        # Vérifier que l'utilisateur est impliqué
        if gift.sender_id != current_user.id and gift.receiver_id != current_user.id:
            raise HTTPException(
                status_code=403,
                detail="Vous n'avez pas accès à ce cadeau"
            )
        
        # Récupérer les infos associées
        sender = db.query(User).filter(User.id == gift.sender_id).first()
        receiver = db.query(User).filter(User.id == gift.receiver_id).first()
        user_bom = db.query(UserBom).filter(UserBom.id == gift.user_bom_id).first()
        boom = user_bom.bom if user_bom else None
        
        # Construire la réponse
        response = {
            "id": gift.id,
            "sender_id": gift.sender_id,
            "sender_name": sender.full_name if sender else f"User {gift.sender_id}",
            "receiver_id": gift.receiver_id,
            "receiver_name": receiver.full_name if receiver else f"User {gift.receiver_id}",
            "user_bom_id": gift.user_bom_id,
            "boom_title": boom.title if boom else "BOOM inconnu",
            "boom_image_url": boom.preview_image if boom else None,
            "message": gift.message,
            "fees": float(gift.fees) if gift.fees else 0.0,
            "status": gift.status.value,
            "is_new_flow": gift.is_new_flow,
            "sent_at": gift.sent_at.isoformat() if gift.sent_at else None,
            "accepted_at": gift.accepted_at.isoformat() if gift.accepted_at else None,
            "expires_at": gift.expires_at.isoformat() if gift.expires_at else None,
            "paid_at": gift.paid_at.isoformat() if gift.paid_at else None,
            "delivered_at": gift.delivered_at.isoformat() if gift.delivered_at else None,
            "failed_at": gift.failed_at.isoformat() if gift.failed_at else None,
            "transaction_reference": gift.transaction_reference
        }
        
        # Détails financiers pour new flow
        if gift.is_new_flow:
            response["financial_details"] = {
                "gross_amount": float(gift.gross_amount) if gift.gross_amount else None,
                "fee_amount": float(gift.fee_amount) if gift.fee_amount else None,
                "net_amount": float(gift.net_amount) if gift.net_amount else None,
                "wallet_transaction_ids": gift.wallet_transaction_ids or []
            }
        
        # Métriques sociales si BOOM disponible
        if boom:
            social_value = float(boom.social_value) if boom.social_value is not None else None
            current_market_value = (
                float(boom.current_social_value)
                if getattr(boom, "current_social_value", None) is not None
                else None
            )
            share_count = boom.share_count or 0
            interaction_count = boom.interaction_count or 0

            response["social_metrics"] = {
                "social_value": social_value,
                "current_market_value": current_market_value,
                "share_count": share_count,
                "interaction_count": interaction_count,
                # Champs hérités conservés pour compatibilité éventuelle côté client
                "boom_social_value": social_value,
                "boom_share_count": share_count,
                "boom_interaction_count": interaction_count
            }
        
        return response
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Erreur récupération détails cadeau: {str(e)}"
        )

File: frontend/src/services/gift.ts
import api from './api';

export interface GiftRequest {
  receiver_phone: string;
  bom_id: number;
  quantity: number;
  message?: string;
}

export interface GiftResponse {
  id: number;
  sender_id: number;
  receiver_id: number;
  is_new_flow: boolean;
  user_bom_id: number;
  message?: string;
  status: 'sent' | 'accepted' | 'declined' | 'expired';
  sent_at: string;
  accepted_at?: string;
  expires_at?: string;
  sender_name: string;
  receiver_name: string;
  bom_title: string;
  bom_image_url?: string;
}

export interface GiftDetailsResponse {
  id: number;
  sender_id: number;
  sender_name?: string;
  receiver_id: number;
  receiver_name?: string;
  user_bom_id: number;
  boom_title?: string;
  boom_image_url?: string | null;
  message?: string | null;
  fees?: number | null;
  status: string;
  is_new_flow: boolean;
  sent_at?: string | null;
  accepted_at?: string | null;
  expires_at?: string | null;
  paid_at?: string | null;
  delivered_at?: string | null;
  failed_at?: string | null;
  transaction_reference?: string | null;
  financial_details?: GiftFinancialBlock;
  social_metrics?: GiftSocialBlock;
}

export interface GiftActionRequest {
  gift_id: number;
  action: 'accepted' | 'declined';
}

export interface GiftActionResponse {
  message: string;
  gift_id: number;
}

export interface GiftTimeline {
  sent_at?: string | null;
  paid_at?: string | null;
  delivered_at?: string | null;
  accepted_at?: string | null;
  declined_at?: string | null;
  expires_at?: string | null;
}

export interface GiftPeopleBlock {
  sender: { id: number; name: string };
  receiver: { id: number; name: string };
}

export interface GiftFinancialBlock {
  gross_amount?: number | null;
  fee_amount?: number | null;
  net_amount?: number | null;
  estimated_value?: number | null;
  currency: string;
  transaction_reference?: string | null;
  wallet_transaction_ids?: number[];
}

export interface GiftSocialBlock {
  social_value?: number | null;
  current_market_value?: number | null;
  share_count: number;
  interaction_count: number;
  market_capitalization?: number | null;
  effective_capitalization?: number | null;
  capitalization_units?: number | null;
  redistribution_pool?: number | null;
}

export interface GiftInboxEntry {
  id: number;
  status: string;
  status_label: string;
  status_tone: 'info' | 'success' | 'danger' | 'muted';
  is_new_flow: boolean;
  message?: string | null;
  direction: 'incoming' | 'outgoing';
  highlight_pending: boolean;
  quantity: number;
  people: GiftPeopleBlock;
  financial: GiftFinancialBlock;
  social?: GiftSocialBlock | null;
  boom?: {
    id: number;
    title: string;
    preview_image?: string | null;
    collection?: string;
    category?: string;
    animation_url?: string | null;
    rarity?: string | null;
  } | null;
  timeline: GiftTimeline;
  actions: {
    can_accept: boolean;
    can_decline: boolean;
    can_view_details: boolean;
  };
}

export interface GiftInboxSummary {
  pending_count: number;
  received_today: number;
  sent_today: number;
  delivered_count: number;
  new_flow_received: number;
  total_value_received?: number | null;
  total_fees_paid?: number | null;
  last_received_at?: string | null;
  needs_attention: boolean;
}

export interface GiftInboxResponse {
  summary: GiftInboxSummary;
  lists: {
    received: GiftInboxEntry[];
    sent: GiftInboxEntry[];
    pending: GiftInboxEntry[];
  };
}

/**
 * Service de gestion des cadeaux Booms - VERSION RÉELLE
 */
export const giftService = {
  /**
   * Envoyer un Bom comme cadeau à un autre utilisateur
   */
  async sendGift(giftData: GiftRequest): Promise<GiftResponse> {
    try {
      console.log('🎁 [SEND_GIFT] Début - Données:', giftData);
      
      const response = await api.post('/gift/send', giftData);
      
      console.log('✅ [SEND_GIFT] Succès - Réponse:', response.data);
      return response.data;
      
    } catch (error: any) {
      const rawMessage = error.response?.data?.detail || 'Erreur lors de l\'envoi du cadeau';
      let errorMessage = rawMessage;
      if (typeof rawMessage === 'string') {
        const normalized = rawMessage.toLowerCase();
        if (normalized.includes('destinataire') && normalized.includes('actif')) {
          errorMessage =
            'Le destinataire est inactif ou supprimé et ne peut pas recevoir de cadeau pour le moment.\n\n' +
            'Vérifiez que :\n' +
            '• le numéro est correct,\n' +
            '• le compte est bien actif,\n' +
            '• le destinataire n’a pas été supprimé.\n\n' +
            'Si besoin, demandez au destinataire de se reconnecter ou contactez le support.';
        }
      }
      if (typeof errorMessage === 'string' && errorMessage.toLowerCase().includes('destinataire')) {
        console.warn('⚠️ [SEND_GIFT] Envoi refusé:', errorMessage);
      } else {
        console.error('❌ [SEND_GIFT] ERREUR:', error);
        console.error('❌ [SEND_GIFT] Message erreur:', errorMessage);
      }
      throw new Error(errorMessage);
    }
  },

  /**
   * Accepter un cadeau reçu
   */
  async acceptGift(giftId: number): Promise<GiftActionResponse> {
    try {
      console.log(`✅ [ACCEPT_GIFT] Début - Gift ID: ${giftId}`);
      
      const requestData = { 
        gift_id: giftId, 
        action: 'accepted' as const
      };
      console.log('✅ [ACCEPT_GIFT] Données envoyées:', requestData);
      
      const response = await api.post('/gift/accept', requestData);
      
      console.log('✅ [ACCEPT_GIFT] Succès - Réponse:', response.data);
      return response.data;
      
    } catch (error: any) {
      console.error('❌ [ACCEPT_GIFT] ERREUR:', error);
      const errorMessage = error.response?.data?.detail || 'Erreur lors de l\'acceptation du cadeau';
      console.error('❌ [ACCEPT_GIFT] Message erreur:', errorMessage);
      throw new Error(errorMessage);
    }
  },

  /**
   * Refuser un cadeau reçu
   */
  async declineGift(giftId: number): Promise<GiftActionResponse> {
    try {
      console.log(`❌ [DECLINE_GIFT] Début - Gift ID: ${giftId}`);
      
      const requestData = { 
        gift_id: giftId, 
        action: 'declined' as const
      };
      console.log('❌ [DECLINE_GIFT] Données envoyées:', requestData);
      
      const response = await api.post('/gift/decline', requestData);
      
      console.log('✅ [DECLINE_GIFT] Succès - Réponse:', response.data);
      return response.data;
      
    } catch (error: any) {
      console.error('❌ [DECLINE_GIFT] ERREUR:', error);
      const errorMessage = error.response?.data?.detail || 'Erreur lors du refus du cadeau';
      console.error('❌ [DECLINE_GIFT] Message erreur:', errorMessage);
      throw new Error(errorMessage);
    }
  },

  /**
   * Vue consolidée "boîte aux cadeaux"
   */
  async getGiftInbox(): Promise<GiftInboxResponse> {
    try {
      console.log('📥 [GIFT_INBOX] Récupération boîte aux cadeaux...');
      const response = await api.get('/gift/inbox');
      console.log('✅ [GIFT_INBOX] Données reçues:', response.data);
      return response.data;
    } catch (error: any) {
      console.error('❌ [GIFT_INBOX] ERREUR:', error);
      const errorMessage = error.response?.data?.detail || 'Erreur lors du chargement des cadeaux';
      throw new Error(errorMessage);
    }
  },

  /**
   * Récupérer l'historique des cadeaux
   */
  async getGiftHistory(giftType: 'sent' | 'received' | 'all' = 'all'): Promise<GiftResponse[]> {
    try {
      console.log(`📨 [GIFT_HISTORY] Début - Type: ${giftType}`);
      
      const response = await api.get(`/gift/history?gift_type=${giftType}`);
      
      console.log(`✅ [GIFT_HISTORY] Succès - ${response.data.length} cadeaux récupérés`);
      console.log('📨 [GIFT_HISTORY] Données:', response.data);
      return response.data;
      
    } catch (error: any) {
      console.error('❌ [GIFT_HISTORY] ERREUR:', error);
      const errorMessage = error.response?.data?.detail || 'Erreur lors de la récupération de l\'historique';
      console.error('❌ [GIFT_HISTORY] Message erreur:', errorMessage);
      
      throw new Error(errorMessage);
    }
  },

  /**
   * Récupérer les cadeaux en attente (reçus non traités)
   */
  async getPendingGifts(): Promise<GiftResponse[]> {
    try {
      console.log('⏳ [PENDING_GIFTS] Début - Récupération cadeaux en attente...');
      
      const gifts = await this.getGiftHistory('received');
      const pendingGifts = gifts.filter(gift => gift.status === 'sent');
      
      console.log(`✅ [PENDING_GIFTS] Succès - ${pendingGifts.length} cadeaux en attente trouvés`);
      return pendingGifts;
      
    } catch (error) {
      console.error('❌ [PENDING_GIFTS] ERREUR:', error);
      return [];
    }
  },

  /**
   * Vérifier si un utilisateur a des cadeaux en attente
   */
  async hasPendingGifts(): Promise<boolean> {
    try {
      console.log('🔍 [HAS_PENDING_GIFTS] Début - Vérification cadeaux en attente...');
      
      const pendingGifts = await this.getPendingGifts();
      const hasPending = pendingGifts.length > 0;
      
      console.log(`✅ [HAS_PENDING_GIFTS] Résultat: ${hasPending}`);
      return hasPending;
      
    } catch (error) {
      console.error('❌ [HAS_PENDING_GIFTS] ERREUR:', error);
      return false;
    }
  },

  async getGiftDetails(giftId: number): Promise<GiftDetailsResponse> {
    const response = await api.get(`/gift/${giftId}/details`);
    return response.data;
  }
};

export default giftService;